docs/pkgs/config.md
# `@boldr/config`

Configuration helper for Node environments. `@boldr/config` combines multiple configuration files
by recursively merging their properties. 

It also allows you to use referenced values from your config.

## Load Precedence

Files are loaded in the following order: 
  
 - `/config/default.(js|json)`
 - `/config/*.(js|json)` (all files in config folder excluding local configs)
 - `/config/env/$NODE_ENV.(js|json)`
 - `/config/env/$NODE_ENV.local.(js|json)`
 - `/config/local.(js|json)`

When there is no `NODE_ENV` set, it defaults to `development`. 

You can adjust configuration directory with `CFG_DIR` environment variable. The path specified will be resolved from the root of your project (`process.cwd()`)  

## Installation

`$ npm install @boldr/config` or `$ yarn add @boldr/config`
 

## Default Config

`@boldr/config` uses plain JS objects as configuration. 

`/.boldr/config/default.js`

```javascript
module.exports = {
  websiteUrl: 'http://localhost:3000',
  server: {
    port: 2121,
    protocol: 'http',
    host: '0.0.0.0',
    prefix: '/api/v1',
    uploadDir: 'public/uploads/tmp'
  },
  logging: {
    level: 'debug',
    file: false,
  },
  token: {
    secret: 'b0ldrk3kwi11s15',
    expiration: 604800000,
  },
  mail: {
    host: 'smtp.example.com',
    user: 'user@user.com',
    password: 'password',
    port: 465,
    ssl: true,
    from: 'hello@boldr.io',
  },
  db: {
    url: 'postgres://postgres:password@localhost:5432/boldr',
  },
  redis: {
    url: 'redis://127.0.0.1:6379/0',
  },

  
  tools: {
    profile: false,
    paths: {
      publicPath: '/static/',
      entry: {
        server: 'src/serverEntry.js',
        client: 'src/clientEntry.js',
      },
      vendor: 'src/vendor.js',
      output: {
        server: 'build/server',
        client: 'build/client',
      },
    },
    vendor: [],
  },
}

```

`/.boldr/config/default.json`

```json
{
  "objectProperty": {
    "path": "/etc/files",
    "anotherProperty": {
      "size": 10
    }
  },
 
  "cache": {
    "expiration": 300
  },
 
  "server": {
    "port": 12345,
    "host": "localhost",
    "version": "1.2.3"
  },
 
  "boolProperty": true,
  "stringProperty": "lol"
 
}
```

## Usage
Load the module into your app, `import config from '@boldr/config';`.
Directly access property on config object. 

Alternatively, you may use `get`/`has` methods.  

```javascript
const config = require('@boldr/config');
console.log(config.cache.expiration);
```

### `config.get(key: string, defaultValue: any): any`

```javascript
const config = require('@boldr/config');
console.log(config.get('cache.expiration')); // prints 300
console.log(config.get('cache.another')); // prints undefined
console.log(config.get('cache.another', 123)); // prints 123
```

### `config.has(key: string): bool`

```javascript
const config = require('@boldr/config');
console.log(config.has('cache.expiration')); // prints true
```

### `config.addOptions(options: object|string, optional: bool)`

This method is used for adding configuration on the fly. You can pass an object with additional 
configuration, or a string path to JS/JSON file, that exports the configuration. 

```javascript
const config = require('@boldr/config');
console.log(config.newKey); // prints undefined
config.addOptions({newKey: 123});
console.log(config.newKey); // prints 123 
```

```javascript
const config = require('@boldr/config');
config.addOptions(__dirname + '/routes'); // load routes.js file exporting routes object
console.log(config.routes); // prints routes object 
```

## Merging configuration

When multiple configuration files has same keys, their values are merged instead of replacement. 

```javascript
const config = require('@boldr/config');
config.addOptions({
  log: {level: 'error', path: 'error.log'}
});
config.addOptions({
  log: {level: 'trace'}
});
console.log(config.log.level); // prints 'trace' 
console.log(config.log.path); // prints 'error.log' 
```


CHANGELOG.md
# Change Log

## [v0.7.0-alpha.1](https://github.com/strues/boldr/tree/v0.7.0-alpha.1) (2017-01-10)
[Full Changelog](https://github.com/strues/boldr/compare/v0.7.0-alpha.0...v0.7.0-alpha.1)

**Fixed bugs:**

- SinglePost view the post image is massive [\#235](https://github.com/strues/boldr/issues/235)
- Admin/Members minified action error [\#234](https://github.com/strues/boldr/issues/234)

**Merged pull requests:**

- 0.7.0-alpha.1 [\#233](https://github.com/strues/boldr/pull/233) ([strues](https://github.com/strues))

## [v0.7.0-alpha.0](https://github.com/strues/boldr/tree/v0.7.0-alpha.0) (2017-01-04)
[Full Changelog](https://github.com/strues/boldr/compare/v0.6.8...v0.7.0-alpha.0)

## [v0.6.8](https://github.com/strues/boldr/tree/v0.6.8) (2016-12-19)
[Full Changelog](https://github.com/strues/boldr/compare/v0.6.6...v0.6.8)

**Merged pull requests:**

- Cleanup and transition API here. [\#232](https://github.com/strues/boldr/pull/232) ([strues](https://github.com/strues))
- Code quality improvements [\#231](https://github.com/strues/boldr/pull/231) ([strues](https://github.com/strues))
- Fixes error with s3 uploading and adds a test [\#230](https://github.com/strues/boldr/pull/230) ([strues](https://github.com/strues))
- API client refactor and bug fixes [\#229](https://github.com/strues/boldr/pull/229) ([strues](https://github.com/strues))
- refactor\(global\): cleanup and standardize api requests. [\#228](https://github.com/strues/boldr/pull/228) ([strues](https://github.com/strues))

## [v0.6.6](https://github.com/strues/boldr/tree/v0.6.6) (2016-11-30)
[Full Changelog](https://github.com/strues/boldr/compare/v0.6.0-alpha.4...v0.6.6)

**Implemented enhancements:**

- Test Coverage [\#161](https://github.com/strues/boldr/issues/161)
- Type checking with Flow [\#160](https://github.com/strues/boldr/issues/160)
- Normalize data in redux stores [\#159](https://github.com/strues/boldr/issues/159)

**Fixed bugs:**

- API requests fired on initial load against 127.0.0.1:80 [\#226](https://github.com/strues/boldr/issues/226)

**Closed issues:**

- Could not get a front to work [\#201](https://github.com/strues/boldr/issues/201)
- Settings and configuration [\#158](https://github.com/strues/boldr/issues/158)
- Icons / SVGs / Graphics [\#100](https://github.com/strues/boldr/issues/100)

**Merged pull requests:**

- v0.6.6 release [\#227](https://github.com/strues/boldr/pull/227) ([strues](https://github.com/strues))
- Separation of API [\#225](https://github.com/strues/boldr/pull/225) ([strues](https://github.com/strues))
- v0.6.5 [\#224](https://github.com/strues/boldr/pull/224) ([strues](https://github.com/strues))
- Update semantic-ui-react to version 0.61.0 🚀 [\#223](https://github.com/strues/boldr/pull/223) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update sw-precache-webpack-plugin to version 0.6.0 🚀 [\#222](https://github.com/strues/boldr/pull/222) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))

## [v0.6.0-alpha.4](https://github.com/strues/boldr/tree/v0.6.0-alpha.4) (2016-11-11)
[Full Changelog](https://github.com/strues/boldr/compare/v0.6.0-alpha.2...v0.6.0-alpha.4)

**Merged pull requests:**

- 0.6.0-alpha.4 release [\#220](https://github.com/strues/boldr/pull/220) ([strues](https://github.com/strues))
- Update postcss-loader to version 1.1.1 🚀 [\#215](https://github.com/strues/boldr/pull/215) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update semantic-ui-react to version 0.60.0 🚀 [\#213](https://github.com/strues/boldr/pull/213) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Develop [\#212](https://github.com/strues/boldr/pull/212) ([strues](https://github.com/strues))
- Update jest-cli to version 17.0.0 🚀 [\#211](https://github.com/strues/boldr/pull/211) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update babel-jest to version 17.0.0 🚀 [\#210](https://github.com/strues/boldr/pull/210) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))

## [v0.6.0-alpha.2](https://github.com/strues/boldr/tree/v0.6.0-alpha.2) (2016-11-07)
[Full Changelog](https://github.com/strues/boldr/compare/v0.6.0-alpha.1...v0.6.0-alpha.2)

**Merged pull requests:**

- 0.6.0-alpha.2 [\#209](https://github.com/strues/boldr/pull/209) ([strues](https://github.com/strues))
- Blocks and Pages [\#208](https://github.com/strues/boldr/pull/208) ([strues](https://github.com/strues))
- alpha.2 -- Fixes happypack randomness [\#203](https://github.com/strues/boldr/pull/203) ([strues](https://github.com/strues))
- chore\(package\): update semantic-ui-react to version 0.58.2 [\#202](https://github.com/strues/boldr/pull/202) ([strues](https://github.com/strues))

## [v0.6.0-alpha.1](https://github.com/strues/boldr/tree/v0.6.0-alpha.1) (2016-11-01)
[Full Changelog](https://github.com/strues/boldr/compare/v0.5.5...v0.6.0-alpha.1)

**Fixed bugs:**

- Token Check [\#138](https://github.com/strues/boldr/issues/138)

**Closed issues:**

- Settings [\#47](https://github.com/strues/boldr/issues/47)

**Merged pull requests:**

- 0.6.0-Alpha.1 [\#200](https://github.com/strues/boldr/pull/200) ([strues](https://github.com/strues))
- Update helmet to version 3.0.0 🚀 [\#196](https://github.com/strues/boldr/pull/196) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update better-npm-run to version 0.0.13 🚀 [\#195](https://github.com/strues/boldr/pull/195) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update webpack-hot-middleware to version 2.13.1 🚀 [\#193](https://github.com/strues/boldr/pull/193) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update semantic-ui-react to version 0.57.0 🚀 [\#192](https://github.com/strues/boldr/pull/192) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update semantic-ui-react to version 0.56.13 🚀 [\#183](https://github.com/strues/boldr/pull/183) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))

web/.boldr/conf.js
/* eslint-disable */
const path = require('path');
const dotenv = require('dotenv');
dotenv.load();

module.exports = {
  env: {
    NODE_ENV: process.env.NODE_ENV,
    BOLDR__SERVER_PORT: process.env.BOLDR__SERVER_PORT,
    BOLDR__DEV_PORT: process.env.BOLDR__DEV_PORT,
    BOLDR__DEBUG: process.env.BOLDR__DEBUG,
  },
  bundle: {
    graphlUri: 'http://localhost:3000/api/v1/graphql',
    verbose: true,
    debug: false,
    cssModules: true,
    wpProfile: false,
    webPath: '/assets/',
    publicDir: path.resolve(__dirname, '../public'),
    assetsDir:  path.resolve(__dirname, '../public/assets'),
    srcDir: path.resolve(__dirname, '../src'),
    client: {
      entry: path.resolve(__dirname, '../src/client/index.js'),
      admin: path.resolve(__dirname, '../src/client/admin.js'),
      bundleDir: path.resolve(__dirname, '../public/assets'),
    },
    server: {
      entry: path.resolve(__dirname, '../src/server/index.js'),
      bundleDir: path.resolve(__dirname, '../lib'),
    },
    vendor: [
      'apollo-client',
      'axios',
      'boldr-ui',
      'boldr-utils',
      'classnames',
      'draft-js',
      'griddle-react',
      'material-ui',
      'normalizr',
      'prop-types',
      'react-apollo',
      'react-dom',
      'react-dropzone',
      'react-helmet',
      'react-redux',
      'react-router-dom',
      'react-router-redux',
      'react-tap-event-plugin',
      'react-transition-group',
      'react',
      'redux-form',
      'redux',
      'reselect',
      'serialize-javascript',
      'styled-components',
      'webfontloader',
    ],
  },
  server: {
    port: process.env.BOLDR__SERVER_PORT,
    host: '127.0.0.1',
    apiPrefix: '/api/v1',
    siteUrl: 'http://localhost:3000',
  },
  logging: {
    level: 'debug',
    file: {
      enable: true,
      dir: 'logs',
      level: 'info',
      filename: 'boldr.api',
    },
  },
  db: {
    url: 'postgres://postgres:password@127.0.0.1:5432/boldr',
    name: 'boldr',
    debug: false,
  },
  redis: {
    url: 'redis://127.0.0.1:6379',
  },
  token: {
    iss: 'boldr',
    aud: '',
    algorithm: 'HS256',
    secret: 'b0ldrk3kwi11s15',
  },
  mail: {
    host: 'smtp.example',
    port: 465,
    ssl: true,
    user: 'hello@boldr.io',
    password: 'password',
    from: 'hello@boldr.io',
  },
  cors: {
    whitelist: ['http://localhost:2121', 'http://localhost:3000'],
  },
};
